File: cortex/setup/detector.py
"""
cortex.setup.detector
---------------------
Auto-detects project stack, CI/CD pipelines, and environment configuration.
Used by ``cortex setup`` to generate project-aware defaults.

EPIC 4: The detector now accepts an optional ``WorkspaceLayout``
to correctly locate ``.github/workflows/`` regardless of layout mode.
"""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cortex.workspace.layout import WorkspaceLayout


@dataclass
class StackInfo:
    """Detected project stack information."""

    language: str = "unknown"
    package_manager: str = "unknown"
    project_name: str = ""
    frameworks: list[str] = field(default_factory=list)
    has_tests: bool = False
    test_command: str = ""
    lint_command: str = ""
    build_command: str = ""
    dev_dependencies: list[str] = field(default_factory=list)


@dataclass
class CIInfo:
    """Detected CI/CD configuration."""

    has_github_actions: bool = False
    workflows: list[str] = field(default_factory=list)
    has_other_ci: bool = False
    ci_type: str = "none"  # github-actions, gitlab-ci, circleci, jenkins, none


@dataclass
class EnvInfo:
    """Detected environment variables relevant to Cortex."""

    has_openai_key: bool = False
    has_anthropic_key: bool = False
    has_ollama: bool = False
    ollama_base_url: str | None = None


@dataclass
class ProjectContext:
    """Complete detected project context."""

    stack: StackInfo = field(default_factory=StackInfo)
    ci: CIInfo = field(default_factory=CIInfo)
    env: EnvInfo = field(default_factory=EnvInfo)
    root: Path = field(default_factory=Path.cwd)
    layout: "WorkspaceLayout | None" = None

    @property
    def project_type(self) -> str:
        """High-level project type label for template selection."""
        s = self.stack
        if s.language == "python":
            return "python"
        if s.language == "javascript" or s.language == "typescript":
            return "node"
        if s.language == "go":
            return "go"
        if s.language == "rust":
            return "rust"
        if s.language == "java" or s.language == "kotlin":
            return "java"
        if s.language == "ruby":
            return "ruby"
        return "generic"


class ProjectDetector:
    """
    Detects the project's language, package manager, frameworks,
    CI/CD setup, and available environment variables.
    """

    def __init__(self, root: Path | None = None):
        self.root = root or Path.cwd()

    def detect(self) -> ProjectContext:
        """Run all detectors and return a combined ProjectContext."""
        from cortex.workspace.layout import WorkspaceLayout

        root = self.root
        try:
            layout = WorkspaceLayout.discover(root)
        except Exception:
            layout = None

        return ProjectContext(
            stack=self._detect_stack(),
            ci=self._detect_ci(),
            env=self._detect_env(),
            root=root,
            layout=layout,
        )

    # ------------------------------------------------------------------
    # Stack detection
    # ------------------------------------------------------------------

    def _detect_stack(self) -> StackInfo:
        info = StackInfo()

        # Check language-specific files in priority order
        detectors = [
            ("python", self._detect_python),
            ("javascript", self._detect_node),
            ("go", self._detect_go),
            ("rust", self._detect_rust),
            ("java", self._detect_java),
            ("ruby", self._detect_ruby),
        ]

        for lang, detector in detectors:
            result = detector()
            if result is not None:
                info.language = lang
                info.package_manager = result.get("package_manager", "unknown")
                info.project_name = result.get("project_name", "")
                info.frameworks = result.get("frameworks", [])
                info.has_tests = result.get("has_tests", False)
                info.test_command = result.get("test_command", "")
                info.lint_command = result.get("lint_command", "")
                info.build_command = result.get("build_command", "")
                info.dev_dependencies = result.get("dev_dependencies", [])
                break

        # Project name fallback
        if not info.project_name:
            info.project_name = self.root.name

        return info

    def _read_json(self, path: Path) -> dict | None:
        try:
            return json.loads(path.read_text(encoding="utf-8"))
        except (FileNotFoundError, json.JSONDecodeError):
            return None

    def _detect_python(self) -> dict | None:
        pyproject = self.root / "pyproject.toml"
        setup_py = self.root / "setup.py"
        requirements = self.root / "requirements.txt"

        if pyproject.exists() or setup_py.exists() or requirements.exists():
            # Check for common Python test files
            has_tests = (
                any(self.root.rglob("test_*.py"))
                or any(self.root.rglob("*_test.py"))
                or (self.root / "tests").is_dir()
            )
            return {
                "package_manager": "pip",
                "project_name": self.root.name,
                "has_tests": has_tests,
                "test_command": "pytest" if has_tests else "python -m unittest",
                "lint_command": "ruff check .",
            }
        return None

    def _detect_go(self) -> dict | None:
        go_mod = self.root / "go.mod"
        if go_mod.exists():
            content = go_mod.read_text(encoding="utf-8")
            module_match = re.search(r"^module\s+(\S+)", content, re.MULTILINE)
            name = module_match.group(1) if module_match else self.root.name
            return {
                "package_manager": "go",
                "project_name": name,
                "has_tests": (self.root / "go_test.go").exists()
                or any(self.root.rglob("*_test.go")),
                "test_command": "go test ./...",
                "lint_command": "golangci-lint run",
            }
        return None

    def _detect_rust(self) -> dict | None:
        # Cargo.toml isn't JSON, just check existence
        cargo_toml = self.root / "Cargo.toml"
        if cargo_toml.exists():
            content = cargo_toml.read_text(encoding="utf-8")
            name_match = re.search(r'name\s*=\s*"([^"]+)"', content)
            return {
                "package_manager": "cargo",
                "project_name": name_match.group(1) if name_match else self.root.name,
                "has_tests": True,  # Rust projects typically have tests
                "test_command": "cargo test",
                "lint_command": "cargo clippy",
                "build_command": "cargo build --release",
            }
        return None

    def _detect_java(self) -> dict | None:
        gradle = self.root / "build.gradle"
        gradle_kts = self.root / "build.gradle.kts"
        maven = self.root / "pom.xml"

        if gradle.exists() or gradle_kts.exists():
            return {
                "package_manager": "gradle",
                "project_name": self.root.name,
                "has_tests": True,
                "test_command": "./gradlew test",
                "lint_command": "./gradlew check",
                "build_command": "./gradlew build",
            }
        if maven.exists():
            return {
                "package_manager": "maven",
                "project_name": self.root.name,
                "has_tests": True,
                "test_command": "mvn test",
                "lint_command": "mvn checkstyle:check",
                "build_command": "mvn package",
            }
        return None

    def _detect_ruby(self) -> dict | None:
        gemfile = self.root / "Gemfile"
        if gemfile.exists():
            content = gemfile.read_text(encoding="utf-8")
            frameworks = []
            if "rails" in content or "gem 'rails'" in content:
                frameworks.append("rails")
            if "rspec" in content or "gem 'rspec'" in content:
                frameworks.append("rspec")
            return {
                "package_manager": "bundler",
                "project_name": self.root.name,
                "frameworks": frameworks,
                "has_tests": (self.root / "spec").exists(),
                "test_command": "bundle exec rspec" if "rspec" in str(frameworks) else "bundle exec rake test",
                "lint_command": "bundle exec rubocop",
            }
        return None

    def _detect_node(self) -> dict | None:
        pkg_path = self.root / "package.json"
        if not pkg_path.exists():
            return None
        pkg = self._read_json(pkg_path)
        if pkg is None:
            return None
        return self._parse_node_json(pkg)

    def _parse_node_json(self, pkg: dict) -> dict:
        deps = {**pkg.get("dependencies", {}), **pkg.get("devDependencies", {})}
        frameworks = []

        # Detect frameworks
        framework_map = {
            "react": "React",
            "next": "Next.js",
            "next.js": "Next.js",
            "vue": "Vue",
            "nuxt": "Nuxt",
            "@angular/core": "Angular",
            "express": "Express",
            "fastify": "Fastify",
            "nestjs": "NestJS",
            "@nestjs/core": "NestJS",
            "svelte": "Svelte",
            "remix": "Remix",
            "@remix-run/node": "Remix",
            "django": "Django",
            "flask": "Flask",
            "fastapi": "FastAPI",
            "bullmq": "BullMQ",
            "bull": "Bull",
            "prisma": "Prisma",
            "sequelize": "Sequelize",
            "typeorm": "TypeORM",
        }
        for dep, fw_name in framework_map.items():
            if dep in deps:
                frameworks.append(fw_name)

        scripts = pkg.get("scripts", {})
        has_tests = bool(scripts.get("test"))
        test_command = scripts.get("test", "echo 'no test script'")
        lint_command = scripts.get("lint", "")
        build_command = scripts.get("build", "")

        return {
            "package_manager": "pnpm" if (self.root / "pnpm-lock.yaml").exists()
            else "yarn" if (self.root / "yarn.lock").exists()
            else "npm",
            "project_name": pkg.get("name", self.root.name),
            "frameworks": frameworks,
            "has_tests": has_tests,
            "test_command": test_command,
            "lint_command": lint_command,
            "build_command": build_command,
            "dev_dependencies": list(pkg.get("devDependencies", {}).keys()),
        }

    # ------------------------------------------------------------------
    # CI/CD detection
    # ------------------------------------------------------------------

    def _detect_ci(self) -> CIInfo:
        info = CIInfo()

        # GitHub Actions
        gh_workflows = self.root / ".github" / "workflows"
        if gh_workflows.exists():
            info.has_github_actions = True
            info.workflows = [
                f.name for f in gh_workflows.iterdir()
                if f.is_file() and f.suffix in (".yml", ".yaml")
            ]

        # GitLab CI
        if (self.root / ".gitlab-ci.yml").exists():
            info.has_other_ci = True
            info.ci_type = "gitlab-ci"

        # CircleCI
        if (self.root / ".circleci" / "config.yml").exists():
            info.has_other_ci = True
            info.ci_type = "circleci"

        # Jenkins
        if (self.root / "Jenkinsfile").exists():
            info.has_other_ci = True
            info.ci_type = "jenkins"

        if not info.has_github_actions and not info.has_other_ci:
            info.ci_type = "none"

        return info

    # ------------------------------------------------------------------
    # Environment detection
    # ------------------------------------------------------------------

    def _detect_env(self) -> EnvInfo:
        return EnvInfo(
            has_openai_key=bool(os.environ.get("OPENAI_API_KEY")),
            has_anthropic_key=bool(os.environ.get("ANTHROPIC_API_KEY")),
            has_ollama=bool(os.environ.get("OLLAMA_BASE_URL")),
            ollama_base_url=os.environ.get("OLLAMA_BASE_URL"),
        )

File: rust/crates/cortex-setup/src/detector.rs
//! Detector de proyecto — porteo de `cortex/setup/detector.py`.

use std::path::Path;

#[derive(Debug, Clone, Default)]
pub struct StackInfo {
    pub language: String,
    pub package_manager: String,
    pub project_name: String,
    pub frameworks: Vec<String>,
    #[allow(dead_code)]
    pub has_tests: bool,
    pub test_command: String,
    pub lint_command: String,
    pub build_command: String,
    #[allow(dead_code)]
    pub dev_dependencies: Vec<String>,
}

#[derive(Debug, Clone, Default)]
pub struct CIInfo {
    #[allow(dead_code)]
    pub has_github_actions: bool,
    #[allow(dead_code)]
    pub workflows: Vec<String>,
    #[allow(dead_code)]
    pub has_other_ci: bool,
    #[allow(dead_code)]
    pub ci_type: String,
}

#[derive(Debug, Clone, Default)]
pub struct EnvInfo {
    pub has_openai_key: bool,
    #[allow(dead_code)]
    pub has_anthropic_key: bool,
    #[allow(dead_code)]
    pub has_ollama: bool,
    #[allow(dead_code)]
    pub ollama_base_url: Option<String>,
}

/// Layout mínimo consumido por los renderers (`layout.is_new_layout`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Layout {
    New,
    Legacy,
}

#[derive(Debug, Clone, Default)]
pub struct ProjectContext {
    pub stack: StackInfo,
    #[allow(dead_code)]
    pub ci: CIInfo,
    pub env: EnvInfo,
    /// None = layout no descubierto (equivalente a `layout=None` en Python).
    pub is_new_layout: Option<bool>,
}

impl ProjectContext {
    /// Detecta el stack/CI/env sobre `root` y el layout por presencia de
    /// `.cortex/workspace.yaml` con `layout_version >= 2` (sufijo fiel para
    /// los campos que consumen los templates; el descubrimiento completo
    /// vive en cortex-app).
    pub fn detect(root: &Path) -> ProjectContext {
        let env = EnvInfo {
            has_openai_key: std::env::var_os("OPENAI_API_KEY").is_some_and(|v| !v.is_empty()),
            has_anthropic_key: std::env::var_os("ANTHROPIC_API_KEY").is_some_and(|v| !v.is_empty()),
            has_ollama: std::env::var_os("OLLAMA_BASE_URL").is_some_and(|v| !v.is_empty()),
            ollama_base_url: std::env::var("OLLAMA_BASE_URL").ok(),
        };
        Self::detect_with(root, &env)
    }

    /// Variante testeable con env explícito (paridad con `_detect_env`).
    pub fn detect_with(root: &Path, env: &EnvInfo) -> ProjectContext {
        let mut ctx = ProjectContext {
            stack: detect_stack(root),
            ci: detect_ci(root),
            env: env.clone(),
            is_new_layout: Some(detect_layout_new(root)),
        };
        if ctx.stack.project_name.is_empty() {
            let name = root
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();
            ctx.stack.project_name = name;
        }
        ctx
    }
}

fn detect_layout_new(root: &Path) -> bool {
    // Espejo del Case 1 de WorkspaceLayout.discover para la raíz dada.
    let ws = root.join(".cortex").join("workspace.yaml");
    if ws.is_file() {
        if let Ok(content) = std::fs::read_to_string(&ws) {
            // parse mínimo de "layout_version:" (fixture controlado)
            for line in content.lines() {
                let line = line.trim();
                if let Some(rest) = line.strip_prefix("layout_version:") {
                    if rest.trim().parse::<i64>().map(|v| v >= 2).unwrap_or(false) {
                        return true;
                    }
                }
            }
        }
    }
    false
}

fn read_json(path: &Path) -> Option<serde_json::Value> {
    let text = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&text).ok()
}

fn walk_contains(
    root: &Path,
    pred: impl Fn(&std::path::PathBuf) -> bool + Copy,
    depth: usize,
) -> bool {
    let Ok(entries) = std::fs::read_dir(root) else {
        return false;
    };
    for entry in entries.flatten() {
        let p = entry.path();
        if pred(&p) {
            return true;
        }
        let descend =
            depth > 0 && p.is_dir() && p.file_name().map(|n| n != ".git").unwrap_or(false);
        if descend && walk_contains(&p, pred, depth - 1) {
            return true;
        }
    }
    false
}

fn detect_stack(root: &Path) -> StackInfo {
    let mut info = StackInfo::default();

    // Orden de prioridad de Python: python, javascript, go, rust, java, ruby.
    let order = ["python", "javascript", "go", "rust", "java", "ruby"];
    for lang in order {
        let result = match lang {
            "python" => detect_python(root),
            "javascript" => detect_node(root),
            "go" => detect_go(root),
            "rust" => detect_rust(root),
            "java" => detect_java(root),
            _ => detect_ruby(root),
        };
        if let Some(result) = result {
            info.language = lang.to_string();
            info.package_manager = result
                .get("package_manager")
                .and_then(|v| v.as_str())
                .unwrap_or("unknown")
                .into();
            info.project_name = result
                .get("project_name")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .into();
            info.frameworks = result
                .get("frameworks")
                .and_then(|v| v.as_array())
                .map(|a| {
                    a.iter()
                        .filter_map(|x| x.as_str().map(str::to_string))
                        .collect()
                })
                .unwrap_or_default();
            info.has_tests = result
                .get("has_tests")
                .and_then(|v| v.as_bool())
                .unwrap_or(false);
            info.test_command = result
                .get("test_command")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .into();
            info.lint_command = result
                .get("lint_command")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .into();
            info.build_command = result
                .get("build_command")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .into();
            info.dev_dependencies = result
                .get("dev_dependencies")
                .and_then(|v| v.as_array())
                .map(|a| {
                    a.iter()
                        .filter_map(|x| x.as_str().map(str::to_string))
                        .collect()
                })
                .unwrap_or_default();
            break;
        }
    }
    info
}

fn map_of(
    pairs: Vec<(&str, serde_json::Value)>,
) -> Option<serde_json::Map<String, serde_json::Value>> {
    let mut m = serde_json::Map::new();
    for (k, v) in pairs {
        m.insert(k.to_string(), v);
    }
    Some(m)
}

fn s(v: &str) -> serde_json::Value {
    serde_json::Value::String(v.into())
}

fn root_name(root: &Path) -> String {
    root.file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_default()
}

fn detect_python(root: &Path) -> Option<serde_json::Map<String, serde_json::Value>> {
    let pyproject = root.join("pyproject.toml");
    let setup_py = root.join("setup.py");
    let requirements = root.join("requirements.txt");
    if pyproject.exists() || setup_py.exists() || requirements.exists() {
        let has_tests = walk_contains(
            root,
            |p| {
                p.is_file()
                    && p.file_name()
                        .map(|n| {
                            let n = n.to_string_lossy();
                            n.starts_with("test_") && n.ends_with(".py")
                                || (n.ends_with("_test.py"))
                        })
                        .unwrap_or(false)
            },
            8,
        ) || root.join("tests").is_dir();
        map_of(vec![
            ("package_manager", s("pip")),
            ("project_name", s(&root_name(root))),
            ("has_tests", serde_json::Value::Bool(has_tests)),
            (
                "test_command",
                s(if has_tests {
                    "pytest"
                } else {
                    "python -m unittest"
                }),
            ),
            ("lint_command", s("ruff check .")),
        ])
    } else {
        None
    }
}

fn detect_go(root: &Path) -> Option<serde_json::Map<String, serde_json::Value>> {
    let go_mod = root.join("go.mod");
    let content = std::fs::read_to_string(&go_mod).ok()?;
    let name = content
        .lines()
        .find_map(|l| l.strip_prefix("module "))
        .map(|m| m.split_whitespace().next().unwrap_or("").to_string())
        .filter(|m| !m.is_empty())
        .unwrap_or_default();
    let has_tests = root.join("go_test.go").exists()
        || walk_contains(
            root,
            |p| {
                p.is_file()
                    && p.file_name()
                        .map(|n| n.to_string_lossy().ends_with("_test.go"))
                        .unwrap_or(false)
            },
            8,
        );
    map_of(vec![
        ("package_manager", s("go")),
        ("project_name", s(&name)),
        ("has_tests", serde_json::Value::Bool(has_tests)),
        ("test_command", s("go test ./...")),
        ("lint_command", s("golangci-lint run")),
    ])
}

fn detect_rust(root: &Path) -> Option<serde_json::Map<String, serde_json::Value>> {
    let cargo_toml = root.join("Cargo.toml");
    let content = std::fs::read_to_string(cargo_toml).ok()?;
    // re.search(r'name\s*=\s*"([^"]+)"')
    let name = find_toml_name(&content).unwrap_or_default();
    map_of(vec![
        ("package_manager", s("cargo")),
        ("project_name", s(&name)),
        ("has_tests", serde_json::Value::Bool(true)),
        ("test_command", s("cargo test")),
        ("lint_command", s("cargo clippy")),
        ("build_command", s("cargo build --release")),
    ])
}

/// Primera aparición de `name = "..."` (semántica de la regex original).
pub(crate) fn find_toml_name(content: &str) -> Option<String> {
    for line in content.lines() {
        let trimmed = line.trim_start();
        if let Some(rest) = trimmed.strip_prefix("name") {
            let rest = rest.trim_start();
            if let Some(rest) = rest.strip_prefix('=') {
                let rest = rest.trim_start();
                if let Some(rest) = rest.strip_prefix('"') {
                    if let Some(end) = rest.find('"') {
                        return Some(rest[..end].to_string());
                    }
                }
            }
        }
    }
    None
}

fn detect_java(root: &Path) -> Option<serde_json::Map<String, serde_json::Value>> {
    if root.join("build.gradle").exists() || root.join("build.gradle.kts").exists() {
        return map_of(vec![
            ("package_manager", s("gradle")),
            ("project_name", s(&root_name(root))),
            ("has_tests", serde_json::Value::Bool(true)),
            ("test_command", s("./gradlew test")),
            ("lint_command", s("./gradlew check")),
            ("build_command", s("./gradlew build")),
        ]);
    }
    if root.join("pom.xml").exists() {
        return map_of(vec![
            ("package_manager", s("maven")),
            ("project_name", s(&root_name(root))),
            ("has_tests", serde_json::Value::Bool(true)),
            ("test_command", s("mvn test")),
            ("lint_command", s("mvn checkstyle:check")),
            ("build_command", s("mvn package")),
        ]);
    }
    None
}

fn detect_ruby(root: &Path) -> Option<serde_json::Map<String, serde_json::Value>> {
    let content = std::fs::read_to_string(root.join("Gemfile")).ok()?;
    let mut frameworks: Vec<String> = Vec::new();
    if content.contains("rails") || content.contains("gem 'rails'") {
        frameworks.push("rails".into());
    }
    if content.contains("rspec") || content.contains("gem 'rspec'") {
        frameworks.push("rspec".into());
    }
    let has_tests = root.join("spec").exists();
    let test_command = if frameworks.iter().any(|f| f == "rspec") {
        "bundle exec rspec"
    } else {
        "bundle exec rake test"
    };
    map_of(vec![
        ("package_manager", s("bundler")),
        ("project_name", s(&root_name(root))),
        (
            "frameworks",
            serde_json::Value::Array(frameworks.into_iter().map(|f| s(&f)).collect()),
        ),
        ("has_tests", serde_json::Value::Bool(has_tests)),
        ("test_command", s(test_command)),
        ("lint_command", s("bundle exec rubocop")),
    ])
}

fn detect_node(root: &Path) -> Option<serde_json::Map<String, serde_json::Value>> {
    let pkg_path = root.join("package.json");
    if !pkg_path.exists() {
        return None;
    }
    let pkg = read_json(&pkg_path)?.as_object()?.clone();

    let deps_value = |k: &str| {
        pkg.get(k)
            .and_then(|v| v.as_object())
            .cloned()
            .unwrap_or_default()
    };
    let mut deps = deps_value("dependencies");
    for (k, v) in deps_value("devDependencies") {
        deps.insert(k, v);
    }

    let framework_map: &[(&str, &str)] = &[
        ("react", "React"),
        ("next", "Next.js"),
        ("next.js", "Next.js"),
        ("vue", "Vue"),
        ("nuxt", "Nuxt"),
        ("@angular/core", "Angular"),
        ("express", "Express"),
        ("fastify", "Fastify"),
        ("nestjs", "NestJS"),
        ("@nestjs/core", "NestJS"),
        ("svelte", "Svelte"),
        ("remix", "Remix"),
        ("@remix-run/node", "Remix"),
        ("django", "Django"),
        ("flask", "Flask"),
        ("fastapi", "FastAPI"),
        ("bullmq", "BullMQ"),
        ("bull", "Bull"),
        ("prisma", "Prisma"),
        ("sequelize", "Sequelize"),
        ("typeorm", "TypeORM"),
    ];
    let frameworks: Vec<String> = framework_map
        .iter()
        .filter(|(dep, _)| deps.contains_key(*dep))
        .map(|(_, fw)| fw.to_string())
        .collect();

    let scripts = pkg
        .get("scripts")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();
    let get_script = |k: &str| scripts.get(k).and_then(|v| v.as_str()).unwrap_or("");
    let has_tests = !get_script("test").is_empty();

    let package_manager = if root.join("pnpm-lock.yaml").exists() {
        "pnpm"
    } else if root.join("yarn.lock").exists() {
        "yarn"
    } else {
        "npm"
    };

    let project_name = pkg
        .get("name")
        .and_then(|v| v.as_str())
        .map(str::to_string)
        .unwrap_or_default();

    let dev_dependencies: Vec<String> = deps_value("devDependencies").keys().cloned().collect();

    map_of(vec![
        ("package_manager", s(package_manager)),
        ("project_name", s(&project_name)),
        (
            "frameworks",
            serde_json::Value::Array(frameworks.into_iter().map(|f| s(&f)).collect()),
        ),
        ("has_tests", serde_json::Value::Bool(has_tests)),
        (
            "test_command",
            s(if get_script("test").is_empty() {
                "echo 'no test script'"
            } else {
                get_script("test")
            }),
        ),
        ("lint_command", s(get_script("lint"))),
        ("build_command", s(get_script("build"))),
        (
            "dev_dependencies",
            serde_json::Value::Array(dev_dependencies.into_iter().map(|d| s(&d)).collect()),
        ),
    ])
}

fn detect_ci(root: &Path) -> CIInfo {
    let mut info = CIInfo::default();
    let gh_workflows = root.join(".github").join("workflows");
    if gh_workflows.is_dir() {
        info.has_github_actions = true;
        if let Ok(entries) = std::fs::read_dir(&gh_workflows) {
            info.workflows = entries
                .flatten()
                .filter(|e| e.path().is_file())
                .filter(|e| {
                    e.path()
                        .extension()
                        .map(|x| x == "yml" || x == "yaml")
                        .unwrap_or(false)
                })
                .map(|e| e.file_name().to_string_lossy().to_string())
                .collect();
        }
    }
    if root.join(".gitlab-ci.yml").exists() {
        info.has_other_ci = true;
        info.ci_type = "gitlab-ci".into();
    }
    if root.join(".circleci").join("config.yml").exists() {
        info.has_other_ci = true;
        info.ci_type = "circleci".into();
    }
    if root.join("Jenkinsfile").exists() {
        info.has_other_ci = true;
        info.ci_type = "jenkins".into();
    }
    if !info.has_github_actions && !info.has_other_ci {
        info.ci_type = "none".into();
    }
    info
}
